File: day21/Day21.hs
{-# LANGUAGE TupleSections #-}
{-# LANGUAGE TypeApplications #-}

import Data.Bifunctor
import Data.Either
import qualified Data.List as L
import qualified Data.Map.Strict as M

import System.Environment

-- Just a primitive lens
data Turn = Turn {_get :: (Int, Int) -> Int, _set :: Int -> (Int, Int) -> (Int, Int)}
data Game = Game
  { _positions :: {-# UNPACK #-} !(Int, Int)
  , _scores :: {-# UNPACK #-} !(Int, Int)
  }
  deriving (Eq, Ord, Show)

type DiceRolls = [[[Int]]]

stdTurns = [Turn fst (first . const), Turn snd (second . const)]

adjust base value amount = (value - 1 + amount) `mod` base + 1

step :: Turn -> Game -> Int -> Game
step (Turn get set) (Game positions scores) die =
  let p = adjust 10 (get positions) die
      s = get scores + p
   in Game (set p positions) (set s scores)

untilScoreReaches :: Int -> DiceRolls -> Game -> (Int, [(Game, Int)])
untilScoreReaches amount rolls start =
  go 0 rolls (cycle stdTurns) (M.singleton start 1) []
 where
  go n _ _ universes acc | M.null universes = (n, acc)
  go n (r : rs) (t : ts) universes acc =
    let (finished, remaining) =
          L.partition (\(Game ps (s1, s2), _) -> s1 >= amount || s2 >= amount)
            . concatMap (\(g, c) -> map ((,c) . step t g . sum) $ sequence r)
            $ M.toList universes
        newUniverses = L.foldl' (\m (g, c) -> M.insertWith (+) g c m) M.empty remaining
     in go (n + length r) rs ts newUniverses (finished ++ acc)

solveGen :: DiceRolls -> Int -> (Int, Int) -> (Int, [(Game, Int)])
solveGen rolls limit ps = untilScoreReaches limit rolls (Game ps (0, 0))

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n l = let (g, rest) = L.splitAt n l in g : chunksOf n rest

-- [[[1], [2], [3]], [[4], [5], [6]], ...]
die100 = chunksOf 3 . map (: []) . cycle $ [1 .. 100]
solve1 = (\(rolls, [(Game _ (s1, s2), _)]) -> rolls * min s1 s2) . solveGen die100 1000

{- Part two -}

-- [[[1, 2, 3], [1, 2, 3], [1, 2, 3]], ...]
dieDirac = chunksOf 3 . repeat $ [1, 2, 3]
solve2 =
  uncurry max
    . bimap sum sum
    . partitionEithers
    . map (\(Game _ (s1, s2), c) -> if s1 > s2 then Left c else Right c)
    . snd
    . solveGen dieDirac 21

solutions =
  [ solve1
  , solve2
  ]

main :: IO ()
main = do
  idx <- read . head <$> getArgs
  print
    . (solutions !! (idx - 1))
    . (\[a, b] -> (a, b))
    . fmap (read @Int . last . words)
    . lines
    =<< getContents

File: day21/Day21.py
from sys import argv, stdin
from collections import defaultdict

text = stdin.read()
lines = text.splitlines()

p1 = int(lines[0].split(" ")[-1])
p2 = int(lines[1].split(" ")[-1])


def adjust(p, amount, target=10):
    p += amount % target
    if p > target:
        return p - target
    else:
        return p


if argv[1] == "1":
    die = 100
    roll_count = 0

    def roll():
        global die
        global roll_count

        roll_count += 1
        die += 1
        if die == 101:
            die = 1
        return die

    ps = [p1, p2]
    scores = [0, 0]
    turn = 0

    while max(scores) < 1000:
        p = adjust(ps[turn], roll() + roll() + roll())
        ps[turn] = p
        scores[turn] += p
        turn = 1 - turn

    print(min(scores) * roll_count)
else:
    universes = {((p1, p2), (0, 0), 0): 1}
    finished = {0: 0, 1: 0}

    while len(universes):
        new_universes = defaultdict(int)

        for (ps, scores, turn), count in universes.items():
            for r1 in [1, 2, 3]:
                for r2 in [1, 2, 3]:
                    for r3 in [1, 2, 3]:
                        p = adjust(ps[turn], r1 + r2 + r3, 10)
                        s = scores[turn] + p

                        nps = list(ps)
                        nps[turn] = p
                        nps = tuple(nps)

                        nscores = list(scores)
                        nscores[turn] = s
                        nscores = tuple(nscores)

                        if max(nscores) >= 21:
                            finished[turn] += count
                        else:
                            new_universes[(nps, nscores, 1 - turn)] += count

        universes = new_universes

    print(max(finished.values()))
